src/mixings/logit_sb_mixing.cc

```diff
@@ -0,0 +1,154 @@
+  auto &rng = bayesmix::Rng::Instance().get();
+  Eigen::VectorXd state_c = state.regression_coeffs;
+  auto priorcast = cast_prior();
+  Eigen::VectorXd prior_mean =
+      bayesmix::to_eigen(priorcast->normal_prior().mean());
+  double prop_var = priorcast->proposal_var();
+  double step = priorcast->step_size();
+  // Loop over clusters
+  for (int h = 0; h < unique_values.size(); h++) {
+    // Compute allocation indicators
+    std::vector<bool> is_curr_clus(n);
+    std::vector<bool> is_prev_clus(n);
+    for (int i = 0; i < n; i++) {
+      if (allocations[i] == h) {
+        is_curr_clus[i] = true;
+      } else if (allocations[i] < h) {
+        is_prev_clus[i] = true;
+      }
+    }
+    // Draw proposed state from its distribution
+    Eigen::VectorXd prop_mean =
+        state_c +
+        step * grad_log_full_cond(state_c, is_curr_clus, is_prev_clus);
+    auto prop_covar = prop_var * Eigen::MatrixXd::Identity(dim, dim);
+    Eigen::VectorXd state_prop =
+        stan::math::multi_normal_rng(prop_mean, prop_covar, rng);
+    // Compute acceptance ratio
+    double prior_ratio = -0.5 * ((state_prop - prior_mean).transpose() *
+                                     precision * (state_prop - prior_mean) -
+                                 (state_c - prior_mean).transpose() *
+                                     precision * (state_c - prior_mean))(0);
+    double like_ratio = log_like(state_prop, is_curr_clus, is_prev_clus) -
+                        log_like(state_c, is_curr_clus, is_prev_clus);
+    double prop_ratio = (-0.5 / prop_var) *
+                        ((state_prop - prop_mean).dot(state_prop - prop_mean) -
+                         (state_c - prop_mean).dot(state_c - prop_mean));
+    double log_accept_ratio = prior_ratio + like_ratio - prop_ratio;
+    // Accept with probability ratio
+    double p = stan::math::uniform_rng(0.0, 1.0, rng);
+    if (p < std::exp(log_accept_ratio)) {
+      state.regression_coeffs = state_prop;
+    }
+  }
+}
+
+void LogitSBMixing::set_state_from_proto(
+    const google::protobuf::Message &state_) {
+  auto &statecast =
+      google::protobuf::internal::down_cast<const bayesmix::LogSBState &>(
+          state_);
+  state.regression_coeffs = bayesmix::to_eigen(statecast.regression_coeffs());
+}
+
+void LogitSBMixing::write_state_to_proto(
+    google::protobuf::Message *out) const {
+  bayesmix::LogSBState state_;
+  bayesmix::to_proto(state.regression_coeffs,
+                     state_.mutable_regression_coeffs());
+  google::protobuf::internal::down_cast<bayesmix::LogSBState *>(out)->CopyFrom(
+      state_);
+}
+
+Eigen::VectorXd LogitSBMixing::get_weights(
+    const Eigen::VectorXd &covariate) const {
+  Eigen::VectorXd eta(num_clusters);
+  Eigen::VectorXd weights(num_clusters);
+  for (int h = 0; h < num_clusters; h++) {
+    eta(h) = covariate.dot(state.regression_coeffs.col(h));
+    weights(h) = sigmoid(eta(h));
+    for (int k = 0; k < h - 1; k++) {
```

Comment:
I’m sure there is a better way to do it avoiding all the nested for loops on k. Probably just precompute all the etas, then store the cumulative products of (1-eta_h)...